### PR TITLE
CORS-2830: Provision AWS Infrastructure with SDK

### PR DESCRIPTION
**Pull request by r4f4**
Provision AWS infrastructure without the use of Terraform.

### CODE REVIEW

File: pkg/destroy/aws/aws.go [hunk 1]
```diff
@@ -290,16 +256,46 @@ func (o *ClusterUninstaller) findUntaggableResources(ctx context.Context, iamCli
 	return resources, nil
 }
 
-// findResourcesToDelete returns the resources that should be deleted.
+// findResourcesToDelete returns the resources that should be deleted
```
Comment (r4f4) [inline pkg/destroy/aws/aws.go:259]: Missed this change. Will drop it before merge.

File: pkg/destroy/aws/aws.go [hunk 2]
```diff
@@ -318,46 +314,45 @@ func (o *ClusterUninstaller) findResourcesToDelete(
...
+		}
+		resources = resources.Union(iamUserResources)
 	}
-	resources = resources.Union(untaggableResources)
 
 	return resources, tagClientsWithResources, utilerrors.NewAggregate(errs)
 }
 
-// findResourcesByTag returns the resources with tags that satisfy the filters.
+// FindResourcesByTag returns the resources with tags that satisfy the filters.
```
Comment (r4f4) [inline pkg/destroy/aws/aws.go:342]: Will drop this change before merge.

File: pkg/destroy/aws/iamhelpers.go [hunk 3]
```diff
@@ -155,10 +155,10 @@ func (search *iamUserSearch) arns(ctx context.Context) ([]string, error) {
 	return arns, err
 }
 
-// findIAMRoles returns the IAM roles for the cluster.
+// FindIAMRoles returns the IAM roles for the cluster.
```
Comment (r4f4) [inline pkg/destroy/aws/iamhelpers.go:158]: Will drop this change before merge.

**Review by r4f4 — COMMENTED** [review]
/hold

**Comment by r4f4** [comment]
The linter complaints about sets can be solved once https://github.com/openshift/installer/pull/7680 merges.

**Comment by r4f4** [comment]
Update: squashed fixup commits.

File: pkg/infrastructure/aws/vpc.go [hunk 4]
```diff
@@ -0,0 +1,1152 @@
...
+					SubnetId:     subnetID,
+				})
+				if err != nil {
+					return nil, fmt.Errorf("failed to associate edge subnet (%s) to route table (%s): %w", aws.StringValue(subnetID), aws.StringValue(table.RouteTableId), err)
+				}
+			}
+			logger.WithFields(logrus.Fields{
+				"subnet":      aws.StringValue(subnetID),
+				"route table": aws.StringValue(table.RouteTableId),
+			}).Infoln(createdOrFoundMsg)
```
Comment (r4f4) [inline pkg/infrastructure/aws/vpc.go:200]: Ops, this was supposed to go together with https://github.com/openshift/installer/pull/7676/commits/afcb170c6209ed09d066f3065bd89ba34437cc32. I'll fix once I have other feedback to work on.

**Comment by r4f4** [comment]
Update: rebased to fix merge conflicts.

**Comment by r4f4** [comment]
Update:
* Rebased to fix merge conflicts
* Added a fix for public-only subnets
* Enable SDK provisioning by default on altinfra image.

**Comment by r4f4** [comment]
Update:
* Rebased on top of current master
* Squashed fixup commits
* Fixed the BYO VPC case so that no VPC resources are created
* Added unit tests for BYO VPC
* Fixed the phz and shared VPC so that the proper route53 client is used
* Fixed linter issues
* Dropped the temporary commit making the aws sdk the default provisioner

**Comment by r4f4** [comment]
Update: I've seen a case where an `InvalidInstanceId.NotFound` error is returned right after a `runInstances` call, causing the wait for the instance to acquire an IP address to abort early. Even though that doesn't cause any problems for the cluster creation, it can cause problems during bootstrap gathering, since the bootstrap IP addr won't have been saved.

**Comment by patrickdillon** [comment]
/retest

Lots of pod pending timeouts

**Comment by patrickdillon** [comment]
This LGTM. Will tag after rebase/squash

**Comment by r4f4** [comment]
Update: rebased on top of current master and squashed commits.
/hold cancel

**Comment by r4f4** [comment]
/label acknowledge-critical-fixes-only
This funtionality is built with a separate build TAG, so it shouldn't affect releases. The only change that could have some kind of impact on standard releases is https://github.com/openshift/installer/pull/7676/commits/8c2c5d5ff563647f353242bda6934d2672db5b66.

**Pull request merged**